### PR TITLE
Fix mssql concurrent insert id

### PR DESCRIPTION
**Pull request by cproensa**
The implementation of db_insert_id() in mssql relied on IDENT_CURRENT
making it not thread safe.
Modify it to use ADOdb native implementation, based on SCOPE_IDENTITY()
which is session independent.

Fixes: #25442

### CODE REVIEW

File: core/database_api.php [hunk 1]
```diff
@@ -451,32 +451,43 @@ function db_result( $p_result, $p_row_index = 0, $p_col_index = 0 ) {
 }
 
 /**
- * return the last inserted id for a specific database table
+ * Return the last inserted ID after a insert statement.
+ * Warning: this function must be used immediately after the insert statement
+ *
+ * This relies in AdoDB to get the entity id when this functionality is available
```
Comment (dregad) [inline core/database_api.php]: on ADOdb

File: core/database_api.php [hunk 2]
```diff
@@ -451,32 +451,43 @@ function db_result( $p_result, $p_row_index = 0, $p_col_index = 0 ) {
...
+ * for the specific driver, and it makes sense in our model.
+ * Natively supported:
+ * - mysqli, using: mysqli_insert_id(connection).
+ * - mssqlnative, using SCOPE_IDENTITY().
+ * Not natively supported:
+ * - pgsql, oracle, using the underlying sequence for the table.
+ *
+ * Since the table is needed for those drivers where a sequence is used, the
+ * $p_table parameter is mandatory to ensure portability.
```
Comment (dregad) [inline core/database_api.php]: I assume you've checked that the function is not called anywhere without the table parameter

File: core/database_api.php [hunk 3]
```diff
@@ -451,32 +451,43 @@ function db_result( $p_result, $p_row_index = 0, $p_col_index = 0 ) {
...
-		}
 	}
-	return $g_db->Insert_ID();
+	if( isset( $t_query ) ) {
+		$t_result = db_query( $t_query );
+		return (int)db_result( $t_result );
+	} else {
+		return $g_db->Insert_ID();
+	}
 }
```
Comment (dregad) [inline core/database_api.php]: Avoid the if() with

``` php
switch( $g_db_functional_type ) {
    ...
    default:
        return $g_db->Insert_ID();
}
$t_result = db_query( $t_query );
return (int)db_result( $t_result );
```

Comment (cproensa) [reply core/database_api.php]: thanks

**Comment by cproensa** [comment]
> I assume you've checked that the function is not called anywhere without the table parameter

Yes, all usages from core call it with the table parameter.

File: core/database_api.php (hunk 1 above)
Comment (dregad) [reply core/database_api.php]: This wasn't fixed... `in AdoDB` -> `on ADOdb`

Comment (dregad) [reply core/database_api.php]: Sorry for picking nits, it's *rely **on***, not *in*
https://dictionary.cambridge.org/dictionary/english/rely-on-upon-sb-sth

Comment (cproensa) [reply core/database_api.php]: > _**on**_, not _in_

sorry i din't notice that

the corrections are welcome, thanks

**Pull request merged**